cache/demux_test.go

```diff
@@ -0,0 +1,171 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+func TestBroadcastBatching(t *testing.T) {
+	tests := []struct {
+		name      string
+		input     []int64
+		wantRevs  []int64
+		wantSizes []int
+	}{
+		{
+			name:      "two groups",
+			input:     []int64{14, 14, 15, 15, 15},
+			wantRevs:  []int64{14, 15},
+			wantSizes: []int{2, 3},
+		},
+		{
+			name:      "single group",
+			input:     []int64{7, 7, 7},
+			wantRevs:  []int64{7},
+			wantSizes: []int{3},
+		},
+		{
+			name:      "all distinct",
+			input:     []int64{1, 2, 3},
+			wantRevs:  []int64{1, 2, 3},
+			wantSizes: []int{1, 1, 1},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			var wg sync.WaitGroup
+			d := newDemux(ctx, &wg, 16, time.Hour)
+			w := newWatcher(len(tt.input)+1, nil)
+			d.Register(w, 0)
+
+			d.Broadcast(eventRevs(tt.input...))
+
+			gotRevs, gotSizes := readBatches(t, w, len(tt.wantRevs))
+
+			if diff := cmp.Diff(tt.wantRevs, gotRevs); diff != "" {
+				t.Fatalf("revision mismatch (-want +got):\n%s", diff)
+			}
+			if diff := cmp.Diff(tt.wantSizes, gotSizes); diff != "" {
+				t.Fatalf("batch size mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestSlowWatcherResync(t *testing.T) {
+	tests := []struct {
+		name             string
+		input            []int64
+		wantInitialRevs  []int64
+		wantInitialSizes []int
+		wantResyncRevs   []int64
+		wantResyncSizes  []int
+	}{
+		{
+			name:             "single event overflow",
+			input:            []int64{1, 2, 3},
+			wantInitialRevs:  []int64{1},
+			wantInitialSizes: []int{1},
+			wantResyncRevs:   []int64{2, 3},
+			wantResyncSizes:  []int{1, 1},
+		},
+		{
+			name:             "multi events batch overflow",
+			input:            []int64{10, 10, 11, 12, 12},
+			wantInitialRevs:  []int64{10},
+			wantInitialSizes: []int{2},
+			wantResyncRevs:   []int64{11, 12},
+			wantResyncSizes:  []int{1, 2},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			var wg sync.WaitGroup
+			d := newDemux(ctx, &wg, 16, 10*time.Millisecond)
+			w := newWatcher(1, nil)
+			d.Register(w, 0)
+
+			d.Broadcast(eventRevs(tt.input...))
+
+			gotInitRevs, gotInitSizes := readBatches(t, w, len(tt.wantInitialRevs))
```

Comment:
This tests will flake. You should not use sleep to trigger resync. Instead don't start `resyncLoop` in background and then call `resync` directly. To avoid creating background goroutine, just create a subconstuctor that doesn't start goroutine.